rfcs/0000-rfc-process.md

```diff
@@ -0,0 +1,127 @@
+bit of a design process and produce a consensus among the Nix community and
+the core team.
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with the ecosystem to understand, and implement.  This should get
+into specifics and corner-cases, and include examples of how the feature is
+used.
+
+## When you need to follow this process
+
+You need to follow this process if you intend to make "substantial" changes to
+the Nix ecosystem. What constitutes a "substantial" change is evolving based
+on community norms, but may include the following.
+
+  - Any semantic or syntactic change to the language that is not a bugfix.
+  - Removing language features
+  - Big restructuring of nixpkgs
+  - Introduction of new interfaces or functions
+  - A controversial change
+
+Some changes do not require an RFC:
+
+  - Adding, updating and removing packages in nixpkgs
+  - Additions only likely to be _noticed by_ other developers-of-nix,
+    invisible to users-of-nix.
+
+If you submit a pull request to implement a new feature without going
+through the RFC process, it may be closed with a polite request to
+submit an RFC first.
+
+## What the process is
+
+In short, to get a major feature added to the Nix ecosystem, one should first
+go throught the RFC process in order to improve the likelyhood of inclusion.
+Here are roughly the steps that one would take:
+
+* Fork the RFC repo https://github.com/NixOS/rfcs
+* Copy `0000-template.md` to `rfcs/0000-my-feature.md` (where
+'my-feature' is descriptive. don't assign an RFC number yet).
+* Fill in the RFC
+* Submit a pull request. The pull request is the time to get review of
+the design from the larger community.
+* Build consensus and integrate feedback. RFCs that have broad support
+are much more likely to make progress than those that don't receive any
+comments.
+
+At this point, the person submitting the RFC should find at least one "buddy"
+that will help him bring the RFC to reality. The goal is to improve the
+chances that the RFC is both desired and likely to be implemented.
+
+Whomever merges the RFC should do the following:
+
+* Assign an id, using the PR number of the RFC pull request. (If the RFC
+  has multiple pull requests associated with it, choose one PR number,
+  preferably the minimal one.)
+* Add the file in the `rfcs/` directory.
+* Create a corresponding issue on the appropriate repo (NixOS/nix, NixOS/nixpkgs, ...).
+* Fill in the remaining metadata in the RFC header, including links for
+  the original pull request(s) and the newly created issue.
+* Commit everything.
+
+Once an RFC becomes active then authors may implement it and submit the
+feature as a pull request to the nix or nixpkgs repo. An 'active' is not a
+rubber stamp, and in particular still does not mean the feature will
+ultimately be merged; it does mean that in principle all the major
+stakeholders have agreed to the feature and are amenable to merging it.
+
```

Comment:
No worries :)

How I see it, the RFC process is more about doing a project proposal. Right now our de-facto "project manager", the person who can make wide decisions is Eelco, and he doesn't have time to extract meaning from reading code on large pull-requests. The goal here is to push some of that decision-making upfront, before all the work is done. It's more costly to the developer but should help Eelco. It doesn't mean that *everything* has to be decided upfront though. I agree that we want to avoid doing waterfall design.

If that's not clear from the existing wording then it should be revised. Maybe "RFC" is a bad term because it's too close to IETF RFCs. How about I rename them to NIP (Nix Improvement Proposals)? I think that was @moretea's idea. (danger: bikeshed ahead :)